libcudacxx/include/cuda/std/__utility/cmp.h

```diff
@@ -33,94 +33,119 @@ _CCCL_PUSH_MACROS
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-#if _CCCL_STD_VER > 2017
 template <class _Tp, class... _Up>
 struct _IsSameAsAny : _Or<_IsSame<_Tp, _Up>...>
 {};
 
 template <class _Tp>
-concept __is_safe_integral_cmp =
-  is_integral_v<_Tp>
-  && !_IsSameAsAny<_Tp,
-                   bool,
-                   char,
-                   char16_t,
-                   char32_t
-#  ifndef _LIBCUDACXX_NO_HAS_CHAR8_T
-                   ,
-                   char8_t
-#  endif
-#  ifndef _LIBCUDACXX_HAS_NO_WIDE_CHARACTERS
-                   ,
-                   wchar_t
-#  endif
-                   >::value;
-
-template <__is_safe_integral_cmp _Tp, __is_safe_integral_cmp _Up>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr bool cmp_equal(_Tp __t, _Up __u) noexcept
+struct __is_safe_integral_cmp
+    : bool_constant<is_integral<_Tp>::value
+                    && !_IsSameAsAny<_Tp,
+                                     bool,
+                                     char,
+                                     char16_t,
+                                     char32_t
+#ifndef _LIBCUDACXX_NO_HAS_CHAR8_T
+                                     ,
+                                     char8_t
+#endif
+#ifndef _LIBCUDACXX_HAS_NO_WIDE_CHARACTERS
+                                     ,
+                                     wchar_t
+#endif
+                                     >::value>
+{};
+
+struct __cmp_equal_impl
 {
-  if constexpr (is_signed_v<_Tp> == is_signed_v<_Up>)
+  template <class _Tp, class _Up, enable_if_t<is_signed<_Tp>::value && is_signed<_Up>::value, int> = 0>
+  static constexpr bool __do_cmp(_Tp __t, _Up __u) noexcept
   {
     return __t == __u;
   }
-  else if constexpr (is_signed_v<_Tp>)
+
+  template <class _Tp, class _Up, enable_if_t<is_signed<_Tp>::value, int> = 0>
```

Comment:
This would be ambiguous if both _Tp and _Up are signed, because we cannot short circuit tempalte instantiations we need to add `!is_signed<_Up>::value`